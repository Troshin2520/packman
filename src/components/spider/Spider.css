.spider {
    display: flex;
    width: 4vw;
    height: 4vw;
    flex-direction: column;
    justify-content: center;
}

.spider-body {
    border-radius: 50% 50% 0 0;
    width: 4vw;
    height: 3vw;
    display: flex;
    justify-content: space-around;
    align-items: center;
}

.spider-blue .spider-body,.spider-blue .spider-claw {
    background-color: blue;
}

.spider-red .spider-body,.spider-red .spider-claw {
    background-color: red;
}

.spider-green .spider-body,.spider-green .spider-claw {
    background-color: green;
}

.spider-yellow .spider-body,.spider-yellow .spider-claw {
    background-color: yellow;
}

.spider-drugged .spider-body,.spider-drugged .spider-claw {
    background-color: lightblue;
}

.spider-bottom {
    display: flex;
}

.spider-eyes {
    display:flex;
    background-color: white;
    width: 1vw;
    height: 1vw;
    border-radius: 100%;
}

.spider-direction-left .spider-bottom {
    justify-content: flex-start;
}

.spider-direction-right .spider-bottom {
    justify-content: flex-end;
}

.spider-direction-left .spider-eyes {
    justify-content: flex-start;
    transition: all 0.3s ease-out;
}

.spider-direction-right .spider-eyes {
    justify-content: flex-end;
    transition: all 0.3s ease-out;
}

.spider-direction-center .spider-eyes {
    justify-content: center;
    transition: all 0.3s ease-out;
}

.spider-direction-middle .spider-eyes {
    align-items: center;
    transition: all 0.3s ease-out;
}

.spider-direction-top .spider-eyes {
    align-items: flex-start;
    transition: all 0.3s ease-out;
}

.spider-direction-bottom .spider-eyes {
    align-items: flex-end;
    transition: all 0.3s ease-out;
}

.spider-direction-right .spider-claws, .spider-direction-left .spider-claws {
    width: 90%;
    transition: all 0.3s ease-out;
}

.spider-drugged .spider-eyes {
    justify-content: center;
    align-items: center;
    transition: all 0.3s ease-out;
}

.spider-drugged  .spider-claws {
    width: 100%;
    justify-content: space-between;
}

.spider-pupils {
    background-color: black;
    width: 0.5vw;
    height: 0.5vw;
    border-radius: 100%;
}

.spider-claws {
    width: 100%;
    display: flex;
    justify-content: space-between;
}

.spider-claw {
    width: 0.6vw;
    height: 0.6vw;
}
