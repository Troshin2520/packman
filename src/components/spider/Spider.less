.spider {
  display: flex;
  width: 4rem;
  height: 4rem;
  flex-direction: column;
  justify-content: center;
  position: absolute;

  & .body {
    border-radius: 50% 50% 0 0;
    width: 4rem;
    height: 3rem;
    display: flex;
    justify-content: space-around;
    align-items: center;
  }

  &.blue .body, &.blue .claw {
    background-color: blue;
  }

  &.red .body, &.red .claw {
    background-color: red;
  }

  &.green .body, &.green .claw {
    background-color: green;
  }

  &.orange .body, &.orange .claw {
    background-color: orange;
  }

  &.drugged .body, &.drugged .claw {
    background-color: lightblue;
  }


  &.eaten .body, &.eaten .claw {
    background-color: transparent;
  }

  & .bottom {
    display: flex;
  }

  & .eyes {
    display: flex;
    background-color: white;
    width: 1rem;
    height: 1rem;
    border-radius: 100%;
  }

  & .direction-left .bottom {
    justify-content: flex-start;
  }

  &.drugged .eyes {
    justify-content: center;
    align-items: center;
    transition: all 0.3s ease-out;
  }

  &.drugged .claws {
    width: 100%;
    justify-content: space-between;
  }

  & .pupils {
    background-color: black;
    width: 0.5rem;
    height: 0.5rem;
    border-radius: 100%;
  }

  & .claws {
    width: 100%;
    display: flex;
    justify-content: space-between;
  }

  & .claw {
    width: 0.6rem;
    height: 0.6rem;
  }

  &.turn-left .bottom {
    justify-content: flex-start;
  }

  &.turn-right .bottom {
    justify-content: flex-end;
  }

  &.turn-right .eyes {
    justify-content: flex-end;
    align-items: center;
    transition: all 0.3s ease-out;
  }

  &.turn-left .eyes {
    justify-content: flex-start;
    align-items: center;
    transition: all 0.3s ease-out;
  }

  &.turn-up .eyes {
    align-items: flex-start;
    justify-content: center;
    transition: all 0.3s ease-out;
  }

  &.turn-down .eyes {
    align-items: flex-end;
    justify-content: center;
    transition: all 0.3s ease-out;
  }

  &.turn-right .claws, &.turn-left .claws {
    width: 90%;
    transition: all 0.3s ease-out;
  }

  &.move {
    &-left, &-right, &-up, &-down {
      animation-direction: normal;
      animation-iteration-count: 1;
      animation-fill-mode: backwards;
      animation-timing-function: linear;
    }

    &-speed-3 {
      animation-duration: 0.6s;
    }

    &-speed-2 {
      animation-duration: 0.5s;
    }

    &-speed-1 {
      animation-duration: 0.4s;
    }

    &-speed-3.drugged,
    &-speed-2.drugged,
    &-speed-1.drugged {
      animation-duration: 1s;
    }

    &-left {
      animation-name: move-left;
    }

    &-right {
      animation-name: move-right;
    }

    &-up {
      animation-name: move-up;
    }

    &-down {
      animation-name: move-down;
    }

    @keyframes move-left {
      to {
        transform: translateX(-4rem);
      }
    }

    @keyframes move-right {
      to {
        transform: translateX(4rem);
      }
    }

    @keyframes move-up {
      to {
        transform: translateY(-4rem);
      }
    }

    @keyframes move-down {
      to {
        transform: translateY(4rem);
      }
    }
  }
}





