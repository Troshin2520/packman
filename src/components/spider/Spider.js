import React, {Component} from 'react';
import ReactDOM from 'react-dom';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import './Spider.less';
import {ACTION_MOVE_SPIDER, ACTION_CHECK_POSITION, BLOCK_SIZE, directions, colors} from '../../constants';

class Spider extends Component {

  constructor(props) {
    super(props);
    this.onAnimationEnd = this.onAnimationEnd.bind(this);
  }

  componentWillUpdate() {
    const node = ReactDOM.findDOMNode(this);
    if (node) {
      node.style.animation = 'none';
      node.offsetHeight;
      node.style.animation = '';
    }
  }

  onAnimationEnd() {
    const {onChangeState, ...params} = this.props;
    this.props.onChangeState(ACTION_MOVE_SPIDER, params);
    this.props.onChangeState(ACTION_CHECK_POSITION, {});
  }


  render() {
    return (<div className={`spider ${this.props.color}
                             ${this.props.drugged > 0 ? 'drugged' : ''}
                             ${this.props.drugged < 0 ? 'eaten' : ''}
                             move-${this.props.move}
                             move-speed-${this.props.speed}
                             turn-${this.props.move}`}
                 style={{top: `${this.props.y * BLOCK_SIZE}rem`,
                         left: `${this.props.x * BLOCK_SIZE}rem`}}
                 onAnimationEnd={this.onAnimationEnd}>
      <div className="body">
        <div className="eyes">
          <div className="pupils"></div>
        </div>
        <div className="eyes">
          <div className="pupils"></div>
        </div>
      </div>
      <div className="bottom">
        <div className="claws">
          <div className="claw"></div>
          <div className="claw"></div>
          <div className="claw"></div>
          <div className="claw"></div>
        </div>
      </div>
    </div>);
  }
}

Spider.propTypes = {
  color: PropTypes.oneOf(colors).isRequired,
  move: PropTypes.oneOf(Object.values(directions)),
  drugged: PropTypes.number,
  x: PropTypes.number.isRequired,
  y: PropTypes.number.isRequired
};

Spider.defaultProps = {
  pursues: false,
  drugged: 0,
  move: directions.up
};


export default connect(
  (state, props) => state.spiders[props.color],
  dispatch => ({
    onChangeState(type, item) {
      dispatch({type: type, payload: item});
    }
  })
)(Spider);