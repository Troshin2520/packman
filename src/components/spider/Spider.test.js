import React from 'react';
import Renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Enzyme, {mount} from 'enzyme';
import Adapter from "enzyme-adapter-react-16";
import Spider from './Spider';

Enzyme.configure({ adapter: new Adapter() });

describe('Spider tests', () => {

  const initialState = {spiders:{
      red: {x: 6, y: 6, move: 'top'},
      green: {x: 6, y: 6, move: 'top'},
      blue: {x: 6, y: 6, move: 'top'},
      orange: {x: 6, y: 6, move: 'top'}
  }};

  const mockStore = configureStore();
  const store = mockStore(initialState);

  it('Spider rendered without crushing', () => {
    const component = Renderer.create(
        <Spider color="red" store={store}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Spider rendered without crushing', () => {
    const component = Renderer.create(
      <Spider color="green" store={store}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Spider rendered without crushing', () => {
    const component = Renderer.create(
      <Spider color="blue" store={store}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Spider rendered without crushing', () => {
    const component = Renderer.create(
      <Spider color="orange" store={store}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });


});