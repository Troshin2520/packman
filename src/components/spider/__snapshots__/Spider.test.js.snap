// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spider tests Spider rendered without crushing 1`] = `
<div
  className="spider red
                             
                             
                             move-up
                             move-speed-undefined
                             turn-up"
  onAnimationEnd={[Function]}
  style={
    Object {
      "left": "24rem",
      "top": "24rem",
    }
  }
>
  <div
    className="body"
  >
    <div
      className="eyes"
    >
      <div
        className="pupils"
      />
    </div>
    <div
      className="eyes"
    >
      <div
        className="pupils"
      />
    </div>
  </div>
  <div
    className="bottom"
  >
    <div
      className="claws"
    >
      <div
        className="claw"
      />
      <div
        className="claw"
      />
      <div
        className="claw"
      />
      <div
        className="claw"
      />
    </div>
  </div>
</div>
`;

exports[`Spider tests Spider rendered without crushing 2`] = `
<div
  className="spider green
                             
                             
                             move-up
                             move-speed-undefined
                             turn-up"
  onAnimationEnd={[Function]}
  style={
    Object {
      "left": "24rem",
      "top": "24rem",
    }
  }
>
  <div
    className="body"
  >
    <div
      className="eyes"
    >
      <div
        className="pupils"
      />
    </div>
    <div
      className="eyes"
    >
      <div
        className="pupils"
      />
    </div>
  </div>
  <div
    className="bottom"
  >
    <div
      className="claws"
    >
      <div
        className="claw"
      />
      <div
        className="claw"
      />
      <div
        className="claw"
      />
      <div
        className="claw"
      />
    </div>
  </div>
</div>
`;

exports[`Spider tests Spider rendered without crushing 3`] = `
<div
  className="spider blue
                             
                             
                             move-up
                             move-speed-undefined
                             turn-up"
  onAnimationEnd={[Function]}
  style={
    Object {
      "left": "24rem",
      "top": "24rem",
    }
  }
>
  <div
    className="body"
  >
    <div
      className="eyes"
    >
      <div
        className="pupils"
      />
    </div>
    <div
      className="eyes"
    >
      <div
        className="pupils"
      />
    </div>
  </div>
  <div
    className="bottom"
  >
    <div
      className="claws"
    >
      <div
        className="claw"
      />
      <div
        className="claw"
      />
      <div
        className="claw"
      />
      <div
        className="claw"
      />
    </div>
  </div>
</div>
`;

exports[`Spider tests Spider rendered without crushing 4`] = `
<div
  className="spider orange
                             drugged
                             
                             move-up
                             move-speed-undefined
                             turn-up"
  onAnimationEnd={[Function]}
  style={
    Object {
      "left": "24rem",
      "top": "24rem",
    }
  }
>
  <div
    className="body"
  >
    <div
      className="eyes"
    >
      <div
        className="pupils"
      />
    </div>
    <div
      className="eyes"
    >
      <div
        className="pupils"
      />
    </div>
  </div>
  <div
    className="bottom"
  >
    <div
      className="claws"
    >
      <div
        className="claw"
      />
      <div
        className="claw"
      />
      <div
        className="claw"
      />
      <div
        className="claw"
      />
    </div>
  </div>
</div>
`;
